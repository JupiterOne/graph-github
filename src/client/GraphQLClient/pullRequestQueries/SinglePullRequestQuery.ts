import utils from '../utils';
import {
  BaseQueryState,
  BuildQuery,
  CursorState,
  IteratePagination,
  ProcessedData,
  PullRequest,
} from '../types';
import { ExecutableQuery } from '../CreateQueryExecutor';
import fragments from '../fragments';

interface QueryState extends BaseQueryState {
  isInitialQuery?: true;
  commits?: CursorState;
  reviews?: CursorState;
  labels?: CursorState;
}

type QueryParams = {
  pullRequestNumber: number;
  repoName: string;
  repoOwner: string;
};

const MAX_REQUESTS_LIMIT = 100;

/**
 * Builds the leanest query possible
 * based on the provided queryState.
 *
 * @param queryParams
 * @param queryState
 */
export const buildQuery: BuildQuery<QueryParams, QueryState> = (
  queryParams,
  queryState,
): ExecutableQuery => {
  const query = `
      query (
        $pullRequestNumber: Int!
        $repoName: String!
        $repoOwner: String!
        $maxLimit: Int!
        ${
          queryState?.isInitialQuery ||
          queryState?.commits?.hasNextPage === true
            ? '$commitsCursor: String'
            : ''
        }
        ${
          queryState?.isInitialQuery ||
          queryState?.reviews?.hasNextPage === true
            ? '$reviewsCursor: String'
            : ''
        }
        ${
          queryState?.isInitialQuery || queryState?.labels?.hasNextPage === true
            ? '$labelsCursor: String'
            : ''
        }
      ) {
          repository(name: $repoName, owner: $repoOwner) {
            pullRequest(number: $pullRequestNumber) {
              ...${pullRequestFields}
              ${
                queryState?.isInitialQuery ||
                queryState?.commits?.hasNextPage === true
                  ? commitsQuery
                  : ''
              }
              ${
                queryState?.isInitialQuery ||
                queryState?.reviews?.hasNextPage === true
                  ? reviewsQuery
                  : ''
              }
              ${
                queryState?.isInitialQuery ||
                queryState?.labels?.hasNextPage === true
                  ? labelsQuery
                  : ''
              } 
            }
          }
          ...${fragments.rateLimit}
      }`;

  return {
    query,
    ...(queryState?.rateLimit && {
      rateLimit: queryState.rateLimit,
    }),
    queryVariables: {
      pullRequestNumber: queryParams.pullRequestNumber,
      repoName: queryParams.repoName,
      repoOwner: queryParams.repoOwner,
      maxLimit: MAX_REQUESTS_LIMIT,
      ...(queryState?.commits?.hasNextPage && {
        commitsCursor: queryState?.commits.endCursor,
      }),
      ...(queryState?.reviews?.hasNextPage && {
        reviewsCursor: queryState?.reviews.endCursor,
      }),
      ...(queryState?.labels?.hasNextPage && {
        labelsCursor: queryState?.labels.endCursor,
      }),
    },
  };
};

const pullRequestFields = `
  on PullRequest {
    additions
    author {
      ...${fragments.teamMemberFields}
    }
    authorAssociation
    baseRefName
    baseRefOid
    baseRepository {
      name
      url
      owner {
        ...${fragments.repositoryOwnerFields}
      }
    }
    body
    changedFiles
    checksUrl
    closed
    closedAt
    # comments  # Maybe someday
    createdAt
    databaseId
    deletions
    editor {
      ...${fragments.userFields}
    }
    # files # Maybe someday
    headRefName
    headRefOid
    headRepository {
      name
      owner {
        ...${fragments.repositoryOwnerFields}
      }
    }
    id
    isDraft
    lastEditedAt
    locked
    mergeCommit {
      ...${fragments.commitFields}
    }
    mergeable
    merged
    mergedAt
    mergedBy {
      ...${fragments.teamMemberFields}
    }
    number
    permalink
    publishedAt
    reviewDecision
    # reviewRequests  # Maybe someday
    state
    # suggestedReviewers  # Maybe someday
    title
    updatedAt
    url
  }`;

const commitsQuery = `
    commits(first: $maxLimit, after: $commitsCursor) {
      totalCount
      nodes {
        commit {
          ...${fragments.commitFields}
        }
      }
      
      pageInfo {
        endCursor
        hasNextPage
      }
    }`;

const reviewsQuery = `
    reviews(first: $maxLimit, after: $reviewsCursor) {
      totalCount
      nodes {
        ...${fragments.reviewFields}
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    }`;

const labelsQuery = `
    labels(first: $maxLimit, after: $labelsCursor) {
      totalCount
      nodes {
        id
        name
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    }`;

export const processResponseData = (
  responseData,
): ProcessedData<QueryState> => {
  const rateLimit = responseData.rateLimit;
  const pullRequest = responseData.repository?.pullRequest;

  return {
    resource: utils.responseToResource(pullRequest),
    queryState: {
      rateLimit: rateLimit,
      commits: pullRequest?.commits?.pageInfo,
      reviews: pullRequest?.reviews?.pageInfo,
      labels: pullRequest?.labels?.pageInfo,
    },
  };
};

/**
 * Handles query pagination for inner resources.
 * Builds final resource while calculating the total cost
 * of the queries.
 * @param queryParams
 * @param iteratee
 * @param execute
 */
const iteratePullRequest: IteratePagination<QueryParams, PullRequest> = async (
  queryParams,
  execute,
  iteratee,
) => {
  let finalResource: PullRequest | undefined = undefined;
  let queryCost = 0;
  let queryState: QueryState = { isInitialQuery: true };
  let paginationComplete = false;

  while (!paginationComplete) {
    const executable = buildQuery(queryParams, queryState);

    const response = await execute(executable);

    const { resource: processedResource, queryState: processedQueryState } =
      processResponseData(response);

    finalResource = joinInnerResources(processedResource, finalResource);
    queryCost += processedQueryState.rateLimit?.cost ?? 0;
    queryState = processedQueryState;
    paginationComplete = isPaginationComplete(processedQueryState);
  }

  if (finalResource) {
    await iteratee(finalResource);
  }

  return {
    totalCost: queryCost,
    limit: queryState?.rateLimit?.limit,
    remaining: queryState?.rateLimit?.remaining,
    resetAt: queryState?.rateLimit?.resetAt,
  };
};

/**
 * Combines the Pull Request resource as
 * pagination occurs.
 * @param newResource
 * @param existingResource
 * @private
 */
const joinInnerResources = (
  newResource: PullRequest,
  existingResource?: PullRequest,
): PullRequest => {
  if (!existingResource) {
    return newResource;
  }

  return {
    ...existingResource,
    commits: existingResource.commits!.concat(newResource?.commits ?? []),
    reviews: existingResource.reviews!.concat(newResource?.reviews ?? []),
    labels: existingResource.labels!.concat(newResource?.labels ?? []),
  };
};

/**
 * Determines if all inner resources have
 * completed pagination.
 * @param queryState
 * @private
 */
const isPaginationComplete = (queryState: QueryState): boolean => {
  return (
    !queryState.commits?.hasNextPage &&
    !queryState.labels?.hasNextPage &&
    !queryState.reviews?.hasNextPage
  );
};

export default { iteratePullRequest };
