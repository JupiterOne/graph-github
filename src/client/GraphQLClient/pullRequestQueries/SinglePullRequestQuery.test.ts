import SinglePullRequestQuery from './SinglePullRequestQuery';
import {
  singleQueryFullResponse,
  singleQueryInnerResourcePaginationComplete,
  singleQueryWithPartialInnerResources,
} from './testResponses';

describe('SinglePullRequestQuery', () => {
  describe('#buildQuery', () => {
    test('first query - no cursors', () => {
      const executableQuery = SinglePullRequestQuery.buildQuery(
        5,
        'musical-palm-tree',
        'J1-Test',
      );

      expect(executableQuery).toMatchSnapshot();
    });

    test('followup query with cursors', () => {
      const queryState = {
        commits: { hasNextPage: false },
        reviews: { hasNextPage: false },
        labels: { hasNextPage: true, endCursor: 'labelsEndCursor' },
      };

      const executableQuery = SinglePullRequestQuery.buildQuery(
        5,
        'musical-palm-tree',
        'J1-Test',
        queryState,
      );

      expect(executableQuery).toMatchSnapshot();
    });
  });

  describe('#processResponseData', () => {
    test('Pulling all data out', () => {
      const result = SinglePullRequestQuery.processResponseData(
        singleQueryFullResponse,
      );

      expect(result).toMatchSnapshot();
    });
    test('Pulls partial data', () => {
      const result = SinglePullRequestQuery.processResponseData(
        singleQueryWithPartialInnerResources,
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('#query', () => {
    test('successfully paginates', async () => {
      // Arrange
      const repo = {
        pullRequestNumber: 4,
        repoName: 'musical-palm-tree',
        repoOwner: 'J1-Test',
      };

      const executor = jest
        .fn()
        .mockResolvedValueOnce(singleQueryFullResponse)
        .mockResolvedValueOnce(singleQueryInnerResourcePaginationComplete);
      const iteratee = jest.fn();

      // Act
      const { rateLimitConsumed } = await SinglePullRequestQuery.query(
        repo,
        iteratee,
        executor,
      );

      // Assert
      expect(rateLimitConsumed).toBe(2);
      expect(executor).toHaveBeenCalledTimes(2);
      expect(executor.mock.calls[0][0].queryVariables).toEqual({
        // no cursors are include on initial call
        maxLimit: 100,
        pullRequestNumber: 4,
        repoName: 'musical-palm-tree',
        repoOwner: 'J1-Test',
      });
      expect(executor.mock.calls[1][0].queryVariables).toEqual({
        // commitsCursor is not included
        labelsCursor: 'Y3Vyc2==',
        reviewsCursor: 'Y3Vyc29yOnYyOpO0M',
        maxLimit: 100,
        pullRequestNumber: 4,
        repoName: 'musical-palm-tree',
        repoOwner: 'J1-Test',
      });

      expect(iteratee).toHaveBeenCalledTimes(1);
      expect(iteratee.mock.calls[0][0]).toMatchSnapshot();
    });
  });
});
