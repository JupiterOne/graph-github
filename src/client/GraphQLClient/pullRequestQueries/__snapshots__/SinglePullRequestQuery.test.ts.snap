// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SinglePullRequestQuery #buildQuery first query - no cursors 1`] = `
Object {
  "query": "
      query (
        $pullRequestNumber: Int!
        $repoName: String!
        $repoOwner: String!
        $maxLimit: Int!
        $commitsCursor: String
        $reviewsCursor: String
        $labelsCursor: String
      ) {
          repository(name: $repoName, owner: $repoOwner) {
            pullRequest(number: $pullRequestNumber) {
              ...
  on PullRequest {
    author {
      ...on User {
    name
    login
  }
    }
    baseRefName
    baseRefOid
    baseRepository {
      name
      owner {
        ...on RepositoryOwner {
          login
        }
      }
    }
    body
    changedFiles
    createdAt
    databaseId
    headRefName
    headRefOid
    headRepository {
      name
      owner {
        ...on RepositoryOwner {
          login
        }
      }
    }
    id
    mergeCommit {
      ...on Commit {
        commitUrl
        oid
      }
      associatedPullRequests(first: 1) {
        nodes {
          id
          number
        }
      }
    }
    merged
    mergedAt
    mergedBy {
      ...on User {
    name
    login
  }
    }
    number
    reviewDecision
    state
    title
    updatedAt
    url
  }
              
    commits(first: $maxLimit, after: $commitsCursor) {
      totalCount
      nodes {
        commit {
          ...on Commit {
    oid
    message
    authoredDate
    author {
      user {
        login # this used to be ...userFields
      }
    }
  }
        }
      }
      
      pageInfo {
        endCursor
        hasNextPage
      }
    }
              
    reviews(first: $maxLimit, after: $reviewsCursor) {
      totalCount
      nodes {
        ...on PullRequestReview {
    commit {
      oid
    }
    author {
      ...on User {
          name
          login
        }
    }
    state
  }
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    }
              
    labels(first: $maxLimit, after: $labelsCursor) {
      totalCount
      nodes {
        name
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    } 
            }
          }
          ...on Query {
    rateLimit {
      limit
      cost
      remaining
      resetAt
    }
  }
      }",
  "queryVariables": Object {
    "maxLimit": 100,
    "pullRequestNumber": 5,
    "repoName": "musical-palm-tree",
    "repoOwner": "J1-Test",
  },
}
`;

exports[`SinglePullRequestQuery #buildQuery followup query with cursors 1`] = `
Object {
  "query": "
      query (
        $pullRequestNumber: Int!
        $repoName: String!
        $repoOwner: String!
        $maxLimit: Int!
        
        
        $labelsCursor: String
      ) {
          repository(name: $repoName, owner: $repoOwner) {
            pullRequest(number: $pullRequestNumber) {
              ...
  on PullRequest {
    author {
      ...on User {
    name
    login
  }
    }
    baseRefName
    baseRefOid
    baseRepository {
      name
      owner {
        ...on RepositoryOwner {
          login
        }
      }
    }
    body
    changedFiles
    createdAt
    databaseId
    headRefName
    headRefOid
    headRepository {
      name
      owner {
        ...on RepositoryOwner {
          login
        }
      }
    }
    id
    mergeCommit {
      ...on Commit {
        commitUrl
        oid
      }
      associatedPullRequests(first: 1) {
        nodes {
          id
          number
        }
      }
    }
    merged
    mergedAt
    mergedBy {
      ...on User {
    name
    login
  }
    }
    number
    reviewDecision
    state
    title
    updatedAt
    url
  }
              
              
              
    labels(first: $maxLimit, after: $labelsCursor) {
      totalCount
      nodes {
        name
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    } 
            }
          }
          ...on Query {
    rateLimit {
      limit
      cost
      remaining
      resetAt
    }
  }
      }",
  "queryVariables": Object {
    "labelsCursor": "labelsEndCursor",
    "maxLimit": 100,
    "pullRequestNumber": 5,
    "repoName": "musical-palm-tree",
    "repoOwner": "J1-Test",
  },
}
`;

exports[`SinglePullRequestQuery #buildQuery followup query with partial cursors 1`] = `
Object {
  "query": "
      query (
        $pullRequestNumber: Int!
        $repoName: String!
        $repoOwner: String!
        $maxLimit: Int!
        
        
        $labelsCursor: String
      ) {
          repository(name: $repoName, owner: $repoOwner) {
            pullRequest(number: $pullRequestNumber) {
              ...
  on PullRequest {
    author {
      ...on User {
    name
    login
  }
    }
    baseRefName
    baseRefOid
    baseRepository {
      name
      owner {
        ...on RepositoryOwner {
          login
        }
      }
    }
    body
    changedFiles
    createdAt
    databaseId
    headRefName
    headRefOid
    headRepository {
      name
      owner {
        ...on RepositoryOwner {
          login
        }
      }
    }
    id
    mergeCommit {
      ...on Commit {
        commitUrl
        oid
      }
      associatedPullRequests(first: 1) {
        nodes {
          id
          number
        }
      }
    }
    merged
    mergedAt
    mergedBy {
      ...on User {
    name
    login
  }
    }
    number
    reviewDecision
    state
    title
    updatedAt
    url
  }
              
              
              
    labels(first: $maxLimit, after: $labelsCursor) {
      totalCount
      nodes {
        name
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    } 
            }
          }
          ...on Query {
    rateLimit {
      limit
      cost
      remaining
      resetAt
    }
  }
      }",
  "queryVariables": Object {
    "labelsCursor": "Y3Vyc2==",
    "maxLimit": 100,
    "pullRequestNumber": 5,
    "repoName": "musical-palm-tree",
    "repoOwner": "J1-Test",
  },
}
`;

exports[`SinglePullRequestQuery #iteratePullRequest successfully paginates 1`] = `
Object {
  "commits": Array [
    Object {
      "id": "commit1",
    },
  ],
  "id": "MDExOlB1bGxSZXF1ZXN0MTA=",
  "labels": Array [
    Object {
      "id": "label1=",
      "name": "bug",
    },
    Object {
      "id": "label2=",
      "name": "documentation",
    },
    Object {
      "id": "label3=",
      "name": "first-task",
    },
    Object {
      "id": "label4=",
      "name": "feature",
    },
  ],
  "mergeCommit": Object {
    "associatedPullRequest": undefined,
    "id": "C_kwDOIAVVA9oAKDg1NjNlOWIyZDYzMzRiNzUwZjhmN2I2MzA4NTgxODg5NTBkMmFiOTI",
  },
  "reviews": Array [
    Object {
      "id": "review1",
    },
    Object {
      "id": "review2",
    },
    Object {
      "id": "review3==",
    },
  ],
  "title": "Update CHANGELOG.md",
}
`;

exports[`SinglePullRequestQuery #processResponseData Pulling all data out 1`] = `
Object {
  "queryState": Object {
    "commits": Object {
      "endCursor": "MQ",
      "hasNextPage": false,
    },
    "labels": Object {
      "endCursor": "Y3Vyc2==",
      "hasNextPage": true,
    },
    "rateLimit": Object {
      "cost": 1,
      "limit": 5000,
      "remaining": 4998,
      "resetAt": "2022-03-07T22:04:07Z",
    },
    "reviews": Object {
      "endCursor": "Y3Vyc29yOnYyOpO0M",
      "hasNextPage": true,
    },
  },
  "resource": Object {
    "commits": Array [
      Object {
        "id": "commit1",
      },
    ],
    "id": "MDExOlB1bGxSZXF1ZXN0MTA=",
    "labels": Array [
      Object {
        "id": "label1=",
        "name": "bug",
      },
      Object {
        "id": "label2=",
        "name": "documentation",
      },
    ],
    "mergeCommit": Object {
      "associatedPullRequest": Object {
        "id": "PR_kwDOIAVVA84_DshD",
        "number": 2,
        "reviewDecision": null,
        "state": "MERGED",
        "url": "https://github.com/j1-ingest/ubiquitous-umbrella/pull/2",
      },
      "id": "C_kwDOIAVVA9oAKDg1NjNlOWIyZDYzMzRiNzUwZjhmN2I2MzA4NTgxODg5NTBkMmFiOTI",
    },
    "reviews": Array [
      Object {
        "id": "review1",
      },
      Object {
        "id": "review2",
      },
    ],
    "title": "Update CHANGELOG.md",
  },
}
`;

exports[`SinglePullRequestQuery #processResponseData Pulls partial data 1`] = `
Object {
  "queryState": Object {
    "commits": Object {
      "endCursor": "MQ",
      "hasNextPage": false,
    },
    "labels": Object {
      "endCursor": "Y3Vyc2==",
      "hasNextPage": true,
    },
    "rateLimit": Object {
      "cost": 1,
      "limit": 5000,
      "remaining": 4998,
      "resetAt": "2022-03-07T22:04:07Z",
    },
    "reviews": undefined,
  },
  "resource": Object {
    "commits": Array [
      Object {
        "id": "MDY6",
      },
    ],
    "id": "MDExOlB1bGxSZXF1ZXN0MTA=",
    "labels": Array [
      Object {
        "id": "MDU6TGFiZWwxODE=",
        "name": "bug",
      },
      Object {
        "id": "MDU6TGFiZWwxODI=",
        "name": "documentation",
      },
    ],
    "reviews": Array [],
    "title": "singleQueryWithPartialInnerResources",
  },
}
`;
